stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -59,80 +451,75 @@ namespace stan {
-        for (int j = 0; j < m2.cols(); j++) {
-          typename Eigen::Matrix<T2, R2, C2>::ConstColXpr ccol(m2.col(j));
-          if (j == 0) {
-            if (i == 0) {
-              result(i, j) = var(new dot_product_vari<T1, T2>(crow, ccol));
-            } else {
-              dot_product_vari<T1, T2> *v2
-                = static_cast<dot_product_vari<T1, T2>*>(result(0, j).vi_);
-              result(i, j)
-                = var(new dot_product_vari<T1, T2>(crow, ccol, NULL, v2));
-            }
-          } else {
-            if (i == 0) {
-              dot_product_vari<T1, T2> *v1
-                = static_cast<dot_product_vari<T1, T2>*>(result(i, 0).vi_);
-              result(i, j)
-                = var(new dot_product_vari<T1, T2>(crow, ccol, v1, NULL));
-            } else /* if (i != 0 && j != 0) */ {
-              dot_product_vari<T1, T2> *v1
-                = static_cast<dot_product_vari<T1, T2>*>(result(i, 0).vi_);
-              dot_product_vari<T1, T2> *v2
-                = static_cast<dot_product_vari<T1, T2>*>(result(0, j).vi_);
-              result(i, j)
-                = var(new dot_product_vari<T1, T2>(crow, ccol, v1, v2));
-            }
-          }
+    boost::enable_if_c<boost::is_same<TA, var>::value ||
+                        boost::is_same<TB, var>::value,
+                        Eigen::Matrix<var, RA, CB> >::type
+    multiply(const Eigen::Matrix<TA, RA, CA> &A,
+             const Eigen::Matrix<TB, CA, CB> &B) {
+      stan::math::check_multiplicable("multiply",
+                                                "A", A,
+                                                "B", B);
+      stan::math::check_not_nan("multiply", "A", A);
+      stan::math::check_not_nan("multiply", "B", B);
+
+      // NOTE: this is not a memory leak, this vari is used in the
+      // expression graph to evaluate the adjoint, but is not needed
+      // for the returned matrix.  Memory will be cleaned up with the
+      // arena allocator.
+      multiply_mat_vari<TA, RA, CA, TB, CB> *baseVari
+        = new multiply_mat_vari<TA, RA, CA, TB, CB>(A, B);
+      Eigen::Matrix<var, RA, CB> AB_v(A.rows(), B.cols());
+      for (size_type i = 0; i < AB_v.size(); ++i) {
+          AB_v.coeffRef(i).vi_ = baseVari->variRefAB_[i];
         }
-      }
-      return result;
+      return AB_v;
     }
 
     /**
-     * Return the scalar product of the specified row vector and
-     * specified column vector.  The return is the same as the dot
-     * product.  The two vectors must be the same size.
-     * @param[in] rv Row vector.
-     * @param[in] v Column vector.
-     * @return Scalar result of multiplying row vector by column vector.
-     * @throw std::domain_error if rv and v are not the same size
+     * Return the scalar product of two vectors.
+     * @tparam TA scalar type row vector A 
+     * @tparam CA Columns matrix A
+     * @tparam TB scalar type vector B 
+     * @param[in] A Row vector.
```

Comment:
Very much appreciated! This is my first real PR in awhile so I'm a little rusty.